### ISSUE TITLE
Increased Compression speeds for high throughput data

### ISSUE THREAD

**Issue by alexsanjoseph** [timescale/timescaledb#3077]
Currently the compression policy seems to run be a single-threaded serial process. However, at high throughput, the ingest rates can be significantly more than what a single thread can handle, and hence cause the lots of uncompressed chunks, even though they're all in the queue.

Is there a way to run compression in parallel such that all the chunks which are uncompressed gets compressed in parallel assuming there is sufficient CPU, RAM etc?

Also, might be related, while trying to create multiple independent threads for compression by running multiple sessions manually, then seem to get blocked with a `ShareUpdateExclusiveLock`.

**Comment by WarriorOfWire** [comment]
Came here to make this issue: This is also a consideration for existing tables with a lot of shards. Doing the compression with 1 cursor per hypertable leaves the server bored while there are months of static segments waiting to be compressed.

**Comment by gpernelle** [comment]
Would love to see it implemented as well. Compressing and decompressing faster would be really helpful, especially when it's so easy to spin up a large machine for that job and then downsizing again.

**Comment by fasmit** [comment]
We're running in the same problem. Initial tests with real world ingestion of one data domain, we set a relatively large chunk size of 2 hours, importing 150k rows/sec or ~50MB/sec for chunks of ~300GB. The compression of such a chunk takes 3.5 hours (~25MB/s). It's not possible to do so in parallel with two sessions either, as noted above. 

This means that the compression will never be able to keep up with live imports, and thus necessarily will fall behind more and more, rendering it effectively useless for a real-world setting. We intend to ingest much more than this...

This is not just a "nice to have" feature request: without it, timescaledb compression is essentially unusable (for us).

**Comment by 0xgeert** [comment]
Why does this issues seem stale? As noted above, compression seems near useless for bigger tables as it stands now.

**Comment by Wenqihai** [comment]
We're running in the same problem.  after importing a large amount of history data, it takes a long time to compress a lot of chunks, because it's not possible to compress chunks in parallel,  is there any good solutions now?

**Comment by jvanns** [comment]
Can't deny this would be an awfully useful feature to see implemented!

**Comment by ysmilda** [comment]
For us this would also be a great feature to have!

**Comment by TheUbuntuGuy** [comment]
We have hit a limit on single thread performance on our servers. We have to rate limit the data ingress (by multiple times) to ensure the compression keeps up. It is a huge waste of modern high core-count CPUs.

**Comment by fasmit** [comment]
Fwiw, since this issue is so old, we moved over to Citus where the compression works fine.

**Comment by alexsanjoseph** [comment]
Pretty sad that there has been no response after two years even with so many votes. Like ^, we have moved on to using Influx (which is not the best otherwise)

**Comment by asiayeah** [comment]
I can confirm the issue is still here (v2.13.1).

Some details are available in https://www.timescale.com/forum/t/timescale-only-uses-a-single-core-in-compression/2517

**Comment by svenklemm** [comment]
There are 2 ways that allow parallelization of compression:
1. you can run compress_chunk in parallel on distinct chunks
2. you can use direct compress during ingestion which works with multiple sessions even on the same chunk

**Comment by kvc0** [comment]
Oh @svenklemm I missed this announcement! Pretty neat feature, direct compress ❤️ . I get a lot out of COPY and this would be a boon.

The requirement for global ordering makes it inaccessible for most of my use cases though, as I insert time series into the same tables from several servers. Each writer sends sorted batches, but subsequent batches even from the same reporter can overlap on time. I don’t need hours of overlap - minutes would work, and I have enough memory to accommodate that. Is this overlap something that can be accommodated in the near future?

**Comment by svenklemm** [comment]
yes, the ordering requirement is optional and when you cannot guarantee it you may still use direct compress but need some additional maintenance afterwards to guarantuee ordered chunks
